### ISSUE TITLE
GROUP BY returns 1 invalid result if no rows meet condition 

### ISSUE THREAD

**Issue by silverhawk184** [agershun/alasql#79]
If there is a query that has no results, but contains a GROUP BY, it returns one result regardless, with the data of "undefined"

>console.log(alasql('SELECT Name FROM ? WHERE 1=0 GROUP BY Name',[[{Name:'test'}]]))
>returns: [{Name: 'undefined'}] but should be []

**Comment by silverhawk184** [comment]
Another issue with GROUP BY is that any field in the result that is not in the group by is 'undefined'. MySQL simply returns the data from the first retrieved row.

>alasql('SELECT Type,Name AS Names FROM ? GROUP BY Type',[[{Type:'PT',Name:'Steve'},{Type:'PT',Name:'Dan'},{Type:'FT',Name:'John'}]]);
Returns 
>[{Names: undefined, Type: "PT"},{Names: undefined, Type: "FT"}]

**Comment by agershun** [comment]
Thank you! For the second question you should:
a) add Name to the GROUP BY clause
b) Use FIRST() aggregator

```js
var res = alasql('SELECT Type,Name FROM ? GROUP BY Type, Name',[...);
```
or
```js
var res = alasql('SELECT Type,FIRST(Name) FROM ? GROUP BY Type',[...]);
```

For the first question - sorry, it still in progress...

**Comment by agershun** [comment]
Try this example [in jsFiddle](http://jsfiddle.net/trr2b7qv/2/)

**Comment by silverhawk184** [comment]
Yes, for first question, I am able to get away with this as I already always check for results:
>var res=alasql('SELECT pkey,Name FROM ? WHERE 1=0 GROUP BY Name',[[{Name:'test'}]]));
>if(res.length>0 && typeof res[0].pkey !== 'undefined'){} //use a field that is guaranteed to never be undefined

**Comment by silverhawk184** [comment]
Sorry, just a tweak on your example for others. (It works exactly as I would have expected it to from within mySQL)
>var res = alasql('SELECT Type,FIRST(Name) FROM ? GROUP BY Type',[...]);

returns [[{Type:"foo","FIRST(Name)":"bar"}]]
Make sure you include the as statement.

>var res = alasql('SELECT Type,FIRST(Name) AS Name FROM ? GROUP BY Type',[...]);
will return [[{Type:"foo",Name:"bar"}]]

Is there a way you can test if the fields are requested naked, to automatically apply the FROM() function when used in  a GROUP BY? Maybe in an upcoming version?

**Comment by mathiasrw** [comment]
:+1:

**Comment by agershun** [comment]
**Is there a way you can test if the fields are requested naked, to automatically apply the FROM() function when used in a GROUP BY?**

I think, that the standard SQL reaction is to throw a error like "column is not found". I think, Alasql can check this, but only for SELECT with GROUP BY clause. Plus, it is hard to recognize these "naked" columns in expressions, like:
```sql
    SELECT mycol+1 FROM mytable GROUP BY othercol
```

**Comment by mathiasrw** [comment]
Is this issue resolved?

**Comment by dogma-systems** [comment]
There is news about this bug? Thanks in advance

**Comment by mathiasrw** [comment]
I dont think so...

**Comment by mathiasrw** [comment]
Well... its now returning `[{}]` instead of the original problem - but its still not correct. 

http://jsfiddle.net/t780pnz1/

**Comment by tran4o** [comment]
adding HAVING count(*) > 0 seems to fix this problem

**Comment by mathiasrw** [comment]
Fix released in v4.11.0
